### DISCUSSION TITLE
Deprecation policy

### DISCUSSION THREAD

**Discussion by wjbenfold** [SciTools/iris#4547] [General]
## Background

[Iris](https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_deprecations.html)' current deprecation policy is quite conservative:

> If you need to make a backwards-incompatible change to a public API that has been included in a release (e.g. deleting a method), then you must first deprecate the old behaviour in at least one release, before removing/updating it in the next major release.

It can be quite a long time between major releases, and as such a deprecation can be around for quite a while before it's actually removed.

For contrast, [Numpy's](https://numpy.org/neps/nep-0023-backwards-compatibility.html) deprecation policy is:

> Removal of deprecated functionality:
> - shall be done after at least 2 releases assuming the current 6-monthly release cycle; if that changes, there shall be at least 1 year between deprecation and removal.
> - shall be listed in the release notes of the release where the removal happened.
> - can be done in any minor, but not bugfix, release.

It's worth noting that Numpy consider their own policy to be quite conservative.

We could move to a deprecation policy closer to Numpy's, specifically one that allows us to remove deprecated functionality at minor releases. This would speed up the deprecation process, which keeps the codebase cleaner and thus easier to maintain and use. It also has the consequence of forcing users to update their code to keep working with Iris (which may or may not be desirable).

## Questions

- What speed of deprecation policy is right for Iris?
- How does this speed compare to our current stated (and actual) speeds?
- Do we want to align to another libraries' policy, or have our own?

## How to implement changes

- Check with community / stakeholders
- Update the Deprecation Policy and Code Maintenance pages in our documentation

**Comment by rcomer** [comment]
Seems reasonable to me in general: users should probably aim to address deprecation warnings when they first appear, rather than waiting till the break happens.

I would ask that we could partly tie the speed to the Met Office's parallel suite cycle, i.e. functionality should be deprecated for at least one production environment before it is removed.  In our group we have code that is imported into more than one project, and those projects don't all get updated to the next environment at the same time.  So it's important that the underlying modules work with consecutive environments (preferably without having to write if-loops on the Iris version!)

**Comment by jamesp** [comment]
Above all, I value the [principle of least suprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).  As a user I should look at the new version number and have a good sense of what to expect. If I have to make tons of changes every time there is a minor version bump because the API has changed I'm going to get a bit grumpy and start pinning things.

If you believe in [semver](https://semver.org/) (I know we don't use semver, we use [PEP440](https://www.python.org/dev/peps/pep-0440/#semantic-versioning), but I believe the spirit is the same). 

> Given a version number MAJOR.MINOR.PATCH, increment the:

> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards compatible manner, and
> PATCH version when you make backwards compatible bug fixes.
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

Deprecation is (by definition) not backwards compatible. So would suggest that it must occur at a major release according to semver. If we choose a different model but keep the same version numbering scheme we introduce opportunity for user astonishment.

@wjbenfold:
> It can be quite a long time between major releases, and as such a deprecation can be around for quite a while before it's actually removed.

If we relax the assumption that it is a long time between major releases then we have more opportunity to make breaking changes, but keep the expected version number contract. We're not numpy in many ways, but notably:
- iris has already made it past v1!
- I would argue our API is not as mature as numpy

EDIT: correct PEP440 reference

**Reply by pp-mo** [reply]
> @jamesp Deprecation is (by definition) not backwards compatible. 

I disagree.  The whole point of deprecated API is that we deprecate it (initially) **instead** of removing it.
It's still there + it still functions, so it is backwards-compatible.
But you should _remove_ things (or radically change them) only in a major release.

**Reply by jamesp** [reply]
@pp-mo sorry I was not clear above. We are saying the same thing. I don’t mind when the deprecation gets announced, only when it is implemented. If I read the excerpt from numpy correctly, they will _implement_ deprecations, ie change the api, on minor releases.

**Reply by pp-mo** [reply]
> If I read the excerpt from numpy correctly, they will implement deprecations, ie change the api, on minor releases.

Yes, I see now that you're right -- [they say](https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals)
> Removal of deprecated functionality .. after at least 2 releases ...[or].. at least 1 year ... can be done in any minor, but not bugfix, release.

So ... Iris did aim to apply **_stricter_ rules than this in the past**, whereby we [allowed breaking changes _only_ in major releases](https://scitools.org.uk/iris/docs/v1.13.0/whitepapers/change_management.html#key-principles-you-can-rely-on)

Declaration of interest : Although this was based on @rhattersley ideas, I wrote it, and I'm not still not clear whether everyone else reallly understood + agreed !
The full "backwards compatibility" of those ideas was the thing that fell apart (and was quietly dropped) when we moved to Dask in Iris 2.0 :  The key reason for that was : it would have been hideously complicated to make an Iris version that could work with **_either_** biggus or Dask at the users option, which these "rules" required.
The other big limitation of this approach was, we needed to make a big caveat about breaking changes introduced by updates to dependencies : Which really only makes sense when users can control the whole environment (so it was always a headache for the internal deployed Science environments)

Despite all that, I think that we **_could_** still maintain that we will introduce code-breaking changes (including removing deprecated features) only at major releases, **if we so choose.**  
I think it would mean considering major releases as more frequent and less totally transformative than they have been in the past.  It would also mean using more compatibility controls, like e.g. the [Future options](https://scitools.org.uk/iris/docs/v1.13.0/whitepapers/change_management.html#future-options-iris-future) idea -- which I think we have basically moved away from + there isn't so appetite for this approach in the team anymore, as it gets quite complicated.
But more frequent (and less transformative) 'major' releases isn't necessarily a bad thing.  
In practice, Iris 3 was somewhat an example of that :  Although IIRC we scheduled this as a major release to allow us to change the default behaviour of cube arithmetic (leniency), in practice it was not really that transformative, and practically _could_ have been managed as a minor release (with a Future control).  
However, we did also then find that we were more-or-less forced to drop Python2 support at that point, so it was really useful to just let us do that

**Reply by jamesp** [reply]
@pp-mo 
> I think it would mean considering major releases as more frequent and less totally transformative than they have been in the past. It would also mean using more compatibility controls, like e.g. the Future options idea -- which I think we have basically moved away from + there isn't so appetite for this approach in the team anymore, as it gets quite complicated.
But more frequent (and less transformative) 'major' releases isn't necessarily a bad thing.

Yes, this is what I meant above:
> If we relax the assumption that it is a long time between major releases then we have more opportunity to make breaking changes, but keep the expected version number contract.

Anyway, pretty sure we're thinking along the same lines here!

**Reply by rcomer** [reply]
> The full "backwards compatibility" of those ideas was the thing that fell apart (and was quietly dropped) when we moved to Dask in Iris 2.0 : The key reason for that was : it would have been hideously complicated to make an Iris version that could work with either biggus or Dask at the users option, which these "rules" required.

Wasn't the Biggus --> Dask move a pretty unusual situation?  I think it's fair enough that there can be exceptions but, _where practicable_ there should still be a deprecation pathway.  I'm particularly thinking of #4414, because I very much hope that `merge`, `extract`, etc would be available as standalone functions for a while before the `CubeList` disappears.

**Reply by pp-mo** [reply]
> Wasn't the Biggus --> Dask move a pretty unusual situation?

Maybe, but it showed that we simply **_can't_** promise to always smooth over the big changes.
Of course we are still doing deprecations where possible.  We have just added several for Iris3.2, and we _could_ have removed some things already deprecated, if we had followed the Numpy approach -- but we have not done so at this time.

To be frank, I think we are just less bothered now about avoiding all possible breaking change.
Our previous policy has been heavily driven by MO-internal considerations, and (internal) use of conda is finally relieving a lot of the concerns about that.

**Reply by trexfeathers** [reply]
I would like to see more frequent major releases, but they might not be well received by a sizeable fraction of our user base - those that use Iris without having consciously embraced scientific Python's fast paced development. Not a blocker, but we should be vigilant to the need for helping people transition when we make a major release.

**Reply by jamesp** [reply]
The problem is not the major release, the problem is the breaking changes!
